src/main/groovy/com/twcable/grabbit/jcr/ProtoNodeDecorator.groovy

```diff
@@ -59,21 +59,6 @@ class ProtoNodeDecorator {
         protoProperties.findAll { !(it.name in [JCR_PRIMARYTYPE, JCR_MIXINTYPES]) }
     }
 
-
-    JcrNodeDecorator writeToJcr(@Nonnull Session session) {
```

Comment:
This is getting confusing. 

A ProtoNodeDecorator wraps a ProtoNode, and provides inspections on the ProtoNode. In addition, it knows how to write it self to the JCR.

A JcrNodeDecorator wraps a node already within the JCR, and provides inspections on that node. In addition, it knows how to write it self to a ProtoNode...

If now we have a multi-step process for creating a JcrNode from a Protonode (like making sure an ancestor is checked out before saving the session, and making the node concrete), maybe that means we need something else. Steps being, create the node and add properties, see if ancestor needs to be checked out, checkout if needed, persist (save session). 

```groovy
JcrNodeDecorator writeToJcr(@Nonull Session session) {
  final node = getNode(session)
  final mixinProperty = getMixinProperty()
  if(mixinProperty) {
    transientNode.addMixins(mixinProperty)
  }
  writableProperties.each { transientNode.writeProperty(it) }
  return transientNode.save()
} 

JcrNodeDecorator getNode(session) {
  JcrNodeDecorator.from(JcrUtils.getOrCreateByPath(......))
}

//Previous concrete implementation goes in here :
class ConcreteJcrNodeDecorator implements JcrNodeDecorator {}

class TransientJcrNodeDecorator implements JcrNodeDecorator {

  Node unsavedNode

  TransientJcrNodeDecorator(JCRNode node) {
    if (!node.isNew()) throw new IllegalStateException(...)...
   }

  ConcreteJcrNodeDecorator save() {
    doSave {
     unsavedNode.save()
   }
  }

  void doSave(Closure<Void> doSave) {
    checkoutIfNeeded()
    doSave.call()
    checkinIfNeeded()
  }

  void checkoutIfNeeded() {
    //check versionable, or ancestor versionable
    //if so, checkout
  }
  void writeProperty(...) {}
   ...
  ...
}